FreeRTOS-Plus/Demo/AWS/Fleet_Provisioning_Windows_Simulator/Fleet_Provisioning_With_CSR_Demo/mqtt_operations.c

```diff
@@ -0,0 +1,1069 @@
+                vTaskDelay( pdMS_TO_TICKS( usNextRetryBackOff ) );
+            }
+        }
+    } while( ( xTlsStatus != TLS_TRANSPORT_SUCCESS ) && ( xBackoffAlgStatus == BackoffAlgorithmSuccess ) );
+
+    return xReturnStatus;
+}
+/*-----------------------------------------------------------*/
+
+static bool prvGetNextFreeIndexForOutgoingPublishes( uint8_t * pucIndex )
+{
+    bool xReturnStatus = false;
+    uint8_t ucIndex = 0;
+
+    configASSERT( pxOutgoingPublishPackets != NULL );
+    configASSERT( pucIndex != NULL );
+
+    for( ucIndex = 0; ucIndex < mqttopMAX_OUTGOING_PUBLISHES; ucIndex++ )
+    {
+        /* A free index is marked by invalid packet id. Check if the the index
+         * has a free slot. */
+        if( pxOutgoingPublishPackets[ ucIndex ].usPacketId == mqttopMQTT_PACKET_ID_INVALID )
+        {
+            xReturnStatus = true;
+            break;
+        }
+    }
+
+    /* Copy the available index into the output param. */
+    if( xReturnStatus == true )
+    {
+        *pucIndex = ucIndex;
+    }
+
+    return xReturnStatus;
+}
+/*-----------------------------------------------------------*/
+
+static void prvCleanupOutgoingPublishAt( uint8_t ucIndex )
+{
+    configASSERT( pxOutgoingPublishPackets != NULL );
+    configASSERT( ucIndex < mqttopMAX_OUTGOING_PUBLISHES );
+
+    /* Clear the outgoing publish packet. */
+    ( void ) memset( &( pxOutgoingPublishPackets[ ucIndex ] ),
+                     0x00,
+                     sizeof( pxOutgoingPublishPackets[ ucIndex ] ) );
+}
+/*-----------------------------------------------------------*/
+
+static void prvCleanupOutgoingPublishes( void )
+{
+    configASSERT( pxOutgoingPublishPackets != NULL );
+
+    /* Clean up all the outgoing publish packets. */
+    ( void ) memset( pxOutgoingPublishPackets, 0x00, sizeof( pxOutgoingPublishPackets ) );
+}
+/*-----------------------------------------------------------*/
+
+static void prvCleanupOutgoingPublishWithPacketID( uint16_t usPacketId )
+{
+    uint8_t ucIndex = 0;
+
+    configASSERT( pxOutgoingPublishPackets != NULL );
+    configASSERT( usPacketId != mqttopMQTT_PACKET_ID_INVALID );
+
+    /* Clean up the saved outgoing publish with packet Id equal to usPacketId. */
+    for( ucIndex = 0; ucIndex < mqttopMAX_OUTGOING_PUBLISHES; ucIndex++ )
+    {
+        if( pxOutgoingPublishPackets[ ucIndex ].usPacketId == usPacketId )
+        {
+            prvCleanupOutgoingPublishAt( ucIndex );
+
+            LogDebug( ( "Cleaned up outgoing publish packet with packet id %u.",
+                        usPacketId ) );
+
+            break;
+        }
+    }
+}
+/*-----------------------------------------------------------*/
+
+static void prvMqttCallback( MQTTContext_t * pxMqttContext,
+                             MQTTPacketInfo_t * pxPacketInfo,
+                             MQTTDeserializedInfo_t * pxDeserializedInfo )
+{
+    uint16_t usPacketIdentifier;
+
+    configASSERT( pxMqttContext != NULL );
```

Comment:
What's the point of a callback context if it's unused?

Reply:
These mqtt_operations changes will be addressed in a future PR, where mqtt_demo_helpers is adapted to work with PKCS11.